File: filemaid/cleaners/docker.py
"""Docker image/system cleanup."""
import shutil
import subprocess

from filemaid.cleaners._result import CleanupResult
from filemaid.cleaners._size import parse_size


def can_run() -> bool:
    return shutil.which("docker") is not None


def run(dry_run: bool, config: dict) -> CleanupResult:
    mode = config.get("dev_cleanup", {}).get("docker", {}).get("mode", "safe")
    if mode == "aggressive":
        cmd = ["docker", "system", "prune", "-af", "--volumes"]
    else:
        cmd = ["docker", "image", "prune", "-f"]
    command = " ".join(cmd)

    if dry_run:
        return CleanupResult(
            name="docker",
            status="dry-run",
            detail=f"would run {command}",
            command=command,
        )

    try:
        result = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            timeout=300,
        )
        output = result.stdout.strip()
        reclaimed = "0 B"
        saved = 0
        for line in output.splitlines():
            if "Total reclaimed space:" in line:
                reclaimed = line.split(":", 1)[-1].strip()
                saved = parse_size(reclaimed) or 0
                break
        return CleanupResult(
            name="docker",
            status="ok",
            saved=saved,
            saved_human=reclaimed,
            detail=output or f"ran {command}",
            command=command,
        )
    except Exception as exc:
        return CleanupResult(
            name="docker",
            status="failed",
            detail=str(exc),
            command=command,
        )

File: internal/cleaners/docker.go
package cleaners

import (
	"strings"

	"github.com/logicminds/filemaid/internal/config"
)

var dockerCleaner = Cleaner{
	Name:   "docker",
	CanRun: dockerCanRun,
	Run:    dockerRun,
}

func dockerCanRun() bool {
	_, err := lookPath("docker")
	return err == nil
}

func dockerRun(dryRun bool, cfg *config.Config) CleanupResult {
	mode := "safe"
	if c, ok := cfg.DevCleanup["docker"]; ok && c.Mode != "" {
		mode = c.Mode
	}

	var cmd []string
	if mode == "aggressive" {
		cmd = []string{"docker", "system", "prune", "-af", "--volumes"}
	} else {
		cmd = []string{"docker", "image", "prune", "-f"}
	}
	command := strings.Join(cmd, " ")

	if dryRun {
		return CleanupResult{
			Name:    "docker",
			Status:  "dry-run",
			Detail:  "would run " + command,
			Command: command,
		}
	}

	out, err := runner.Run(cmd[0], cmd[1:]...)
	if err != nil {
		return CleanupResult{
			Name:    "docker",
			Status:  "failed",
			Detail:  err.Error(),
			Command: command,
		}
	}

	output := strings.TrimSpace(string(out))
	reclaimed := "0 B"
	var saved int64
	for _, line := range strings.Split(output, "\n") {
		if strings.Contains(line, "Total reclaimed space:") {
			parts := strings.SplitN(line, ":", 2)
			if len(parts) == 2 {
				reclaimed = strings.TrimSpace(parts[1])
				if s := ParseSize(reclaimed); s != nil {
					saved = *s
				}
			}
			break
		}
	}

	detail := output
	if detail == "" {
		detail = "ran " + command
	}

	return CleanupResult{
		Name:       "docker",
		Status:     "ok",
		Saved:      &saved,
		SavedHuman: reclaimed,
		Detail:     detail,
		Command:    command,
	}
}
